File: app/src/main/java/begovac/rectangles/repository/realm/RealmDataRepository.java
package begovac.rectangles.repository.realm;

import android.support.annotation.Nullable;

import java.util.List;

import begovac.rectangles.model.Data;
import begovac.rectangles.model.Rectangle;
import begovac.rectangles.repository.Repository;
import io.realm.Realm;
import io.realm.RealmList;
import java8.util.function.Function;
import java8.util.stream.Collectors;
import java8.util.stream.StreamSupport;

/** Created by juraj begovac on 26/07/2017. */
public class RealmDataRepository implements Repository<Data> {

  @Nullable
  @Override
  public Data getFirst() {
    RealmData data = Realm.getDefaultInstance().where(RealmData.class).findFirst();

    if (data == null) return null;

    RealmList<RealmRectangle> realmRectangles = data.rectangles;

    List<Rectangle> rectangles =
        StreamSupport.stream(realmRectangles)
            .map(
                new Function<RealmRectangle, Rectangle>() {
                  @Override
                  public Rectangle apply(RealmRectangle realmRectangle) {
                    return Rectangle.create(
                        realmRectangle.getOrdinalNumber(),
                        realmRectangle.getColor(),
                        realmRectangle.getIndexInList());
                  }
                })
            .collect(Collectors.<Rectangle>toList());

    return Data.create(data.rows, data.columns, rectangles);
  }

  @Override
  public void save(Data data) {
    Realm realm = Realm.getDefaultInstance();

    realm.beginTransaction();
    RealmData realmData = realm.createObject(RealmData.class);
    realmData.rows = data.rows();
    realmData.columns = data.columns();

    for (Rectangle r : data.rectangles()) {
      RealmRectangle realmRectangle = realm.createObject(RealmRectangle.class);
      realmRectangle.setOrdinalNumber(r.ordinalNumber());
      realmRectangle.setColor(r.color());
      realmRectangle.setIndexInList(r.indexInList());
      realmData.rectangles.add(realmRectangle);
    }
    realm.commitTransaction();
  }

  @Override
  public void deleteAll() {
    Realm realm = Realm.getDefaultInstance();
    realm.beginTransaction();
    realm.deleteAll();
    realm.commitTransaction();
  }
}

File: app/src/main/kotlin/begovac/rectangles/repository/realm/RealmDataRepository.kt
package begovac.rectangles.repository.realm

import begovac.rectangles.model.Data
import begovac.rectangles.model.Rectangle
import begovac.rectangles.repository.Repository
import io.realm.Realm

/** Created by juraj begovac on 26/07/2017.  */
class RealmDataRepository : Repository<Data> {
  
  override val first: Data?
    get() {
      val data = Realm.getDefaultInstance().where(RealmData::class.java).findFirst() ?: return null
      
      val rectangles = data.rectangles
          .map { Rectangle(it.ordinalNumber, it.color, it.indexInList) }
      
      return Data(data.rows, data.columns, rectangles)
    }
  
  override fun save(item: Data) {
    val realm = Realm.getDefaultInstance()
    
    realm.beginTransaction()
    val realmData = realm.createObject(RealmData::class.java)
    realmData.rows = item.rows
    realmData.columns = item.columns
    item.rectangles
        .forEach {
          val realmRectangle = realm.createObject(RealmRectangle::class.java)
          realmRectangle.ordinalNumber = it.ordinalNumber
          realmRectangle.color = it.color
          realmRectangle.indexInList = it.indexInList
          realmData.rectangles.add(realmRectangle)
        }
    realm.commitTransaction()
  }
  
  override fun deleteAll() {
    val realm = Realm.getDefaultInstance()
    realm.beginTransaction()
    realm.deleteAll()
    realm.commitTransaction()
  }
}
